src/QuickCheckVEngine/RVFI_DII/RVFI.hs

```diff
@@ -73,175 +84,404 @@ type RV_RegIdx = Word8
 -- | Type synonym for a RISCV XLEN-bit word
 type RV_WordXLEN = Word64
 
--- | Type synonym for a RISCV XLEN-bit word
-type RV_WordXLEN_ByteMask = Word8
+data RV_PrivMode = PRV_U | PRV_S | PRV_Reserved | PRV_M deriving (Enum, Show, Eq)
+
+type RV_XL = Word8
 
 -- * Definition of the RISC-V Formal Interface
 
 -- | The 'RVFI_Packet' type captures (a subset of) the RISC-V Formal Interface
-data RVFI_Packet = RVFI_Packet {
--- Metadata
-  rvfi_valid :: Word8
-, rvfi_order :: Word64
-, rvfi_insn  :: Word64
-, rvfi_trap  :: Word8
-, rvfi_halt  :: Word8
-, rvfi_intr  :: Word8
---, rvfi_mode  :: RV_PrivMode -- TODO
---, rvfi_ixl   :: RV_XL       -- TODO
--- Integer Register Read/Write
-, rvfi_rs1_addr  :: RV_RegIdx
-, rvfi_rs2_addr  :: RV_RegIdx
-, rvfi_rs1_rdata :: RV_WordXLEN
-, rvfi_rs2_rdata :: RV_WordXLEN
-, rvfi_rd_addr   :: RV_RegIdx
-, rvfi_rd_wdata  :: RV_WordXLEN
--- Program Counter
-, rvfi_pc_rdata :: RV_WordXLEN
-, rvfi_pc_wdata :: RV_WordXLEN
--- Memory Access
-, rvfi_mem_addr  :: RV_WordXLEN
-, rvfi_mem_rmask :: RV_WordXLEN_ByteMask
-, rvfi_mem_wmask :: RV_WordXLEN_ByteMask
-, rvfi_mem_rdata :: RV_WordXLEN
-, rvfi_mem_wdata :: RV_WordXLEN
--- Further TODOs in the RVFI specification:
--- Control and Status Registers (CSRs)
--- Modelling of Floating-Point State
--- Handling of Speculative Execution
--- Modelling of Virtual Memory
--- Modelling of Atomic Memory Operations
--- Skipping instructions
-}
+data RVFI_Packet = RVFI_Packet
+  { -- Metadata
+    rvfi_valid :: Word8,
+    rvfi_order :: Word64,
+    rvfi_insn :: Word64,
+    rvfi_trap :: Word8,
+    rvfi_halt :: Word8,
+    rvfi_intr :: Word8,
+    rvfi_mode :: Maybe RV_PrivMode,
+    rvfi_ixl :: Maybe RV_XL,
+    -- Program Counter
+    rvfi_pc_rdata :: RV_WordXLEN,
+    rvfi_pc_wdata :: RV_WordXLEN,
+    -- Integer Register Read/Write
+    rvfi_int_data :: Maybe RVFI_IntData,
+    -- Memory Access
+    rvfi_mem_data :: Maybe RVFI_MemAccessData
+    -- Further TODOs in the RVFI specification:
+    -- Control and Status Registers (CSRs)
+    -- Modelling of Floating-Point State
+    -- Handling of Speculative Execution
+    -- Modelling of Virtual Memory
+    -- Modelling of Atomic Memory Operations
+    -- Skipping instructions
+  }
```

Comment:
tiny detail: can we prefer the comma at beginning of the line? :) Makes it easier to add fields to the end (and also generally consistent with the rest of the code)

Reply:
there are a few other places in the file where I'd like that too, but I also don't mind doing it myself after a merge

Reply:
I blame the code formatter (apparently Ormolu). In all other languages hitting the reformat shortcut only reformats the selected lines, but that does not seem to work in the Haskell plugin, so it reformatted the whole file.